src/lib/Types/Core.hs

```diff
@@ -217,6 +217,46 @@ instance GenericB EffectBinder where
   fromB (EffectBinder effs) = LiftB effs
   toB   (LiftB effs) = EffectBinder effs
 
+-- === type classes ===
+
+data ClassDef (n::S) where
+  ClassDef
+    :: SourceName
+    -> [SourceName]       -- method source names
+    -> Nest Binder n1 n2  -- parameters
+    ->   [Type n2]        -- superclasses
+    ->   [MethodType n2]  -- method types
+    -> ClassDef n1
+
+data InstanceDef (n::S) where
+  InstanceDef
+    :: ClassName n1
+    -> Nest PiBinder n1 n2 -- parameters (types and dictionaries)
+    ->   [Type n2]         -- class parameters
+    ->   InstanceBody n2
+    -> InstanceDef n1
+
+data InstanceBody (n::S) =
+  InstanceBody
+    [Atom n]   -- superclasses dicts
+    [Block n]  -- method definitions
+  deriving (Show, Generic)
+
+data MethodType (n::S) =
+  MethodType
+    [Bool]    -- indicates explicit args
+    (Type n)  -- actual method type
+ deriving (Show, Generic)
+
+data DictType (n::S) = DictType SourceName (ClassName n) [Type n]
+     deriving (Show, Generic)
+
+data DictExpr (n::S) =
+   InstanceDict (InstanceName n) [Atom n]
+ | InstantiatedGiven (Atom n) (NonEmpty (Atom n))
```

Comment:
> I feel like this needs more docs.
Yes, agreed

> Why can givens only be referred to when there's at least one argument to apply? It doesn't sound like a true property to me.
This was an attempt to normalize things a bit. You can have a given without arguments indeed, that's the common case! But you have to express it as the `Var` case of `Atom` rather than using a `DictExpr`. If we allowed givens without arguments in `DictExpr` then you could express the same atom as either `Var v` or `DictCon (Given (Var v) [])`. Now, maybe that's actually ok, because we don't make any promises that `DictExpr`s are normalized. wdyt?

Reply:
I think that would be ok, but perhaps this is fine as well, given that our `App`s are non-empty as well.